domain/machine/state/state.go

```diff
@@ -36,64 +36,140 @@ func NewState(factory coredb.TxnRunnerFactory, logger logger.Logger) *State {
-	queryStmt, err := st.Prepare(query, machineNameParam)
+	mName := args.name
+
+	// Prepare query for machine uuid.
+	machineNameParam := machineName{Name: mName}
+	machineUUIDout := machineUUID{}
+	machineUUIDQuery := `SELECT &machineUUID.uuid FROM machine WHERE name = $machineName.name`
+	machineUUIDStmt, err := st.Prepare(machineUUIDQuery, machineNameParam, machineUUIDout)
 	if err != nil {
 		return errors.Trace(err)
 	}
 
-	createMachine := `
+	// Prepare query for creating machine row.
+	createParams := sqlair.M{
+		"machine_uuid":  args.machineUUID,
+		"net_node_uuid": args.netNodeUUID,
+		"name":          mName,
+		"life_id":       life.Alive,
+	}
+	createMachineQuery := `
 INSERT INTO machine (uuid, net_node_uuid, name, life_id)
 VALUES ($M.machine_uuid, $M.net_node_uuid, $M.name, $M.life_id)
 `
-	createMachineStmt, err := st.Prepare(createMachine, machineNameParam)
+	createMachineStmt, err := st.Prepare(createMachineQuery, createParams)
 	if err != nil {
 		return errors.Trace(err)
 	}
 
-	createNode := `INSERT INTO net_node (uuid) VALUES ($M.net_node_uuid)`
-	createNodeStmt, err := st.Prepare(createNode, machineNameParam)
+	// Prepare query for creating net node row.
+	createNodeQuery := `INSERT INTO net_node (uuid) VALUES ($M.net_node_uuid)`
+	createNodeStmt, err := st.Prepare(createNodeQuery, createParams)
 	if err != nil {
 		return errors.Trace(err)
 	}
 
-	createParams := sqlair.M{
-		"machine_uuid":  machineUUID,
-		"net_node_uuid": nodeUUID,
-		"name":          machineName,
-		"life_id":       life.Alive,
+	// Prepare query for associating parent machine.
+	var parentNameParam machineName
+	var associateParentStmt *sqlair.Statement
+	var associateParentParam machineParent
+	if args.parentName != "" {
+		parentNameParam = machineName{Name: args.parentName}
+		associateParentParam = machineParent{MachineUUID: args.machineUUID}
+		associateParentQuery := `
+INSERT INTO machine_parent (machine_uuid, parent_uuid)
+VALUES ($machineParent.machine_uuid, $machineParent.parent_uuid)
+`
+		associateParentStmt, err = st.Prepare(associateParentQuery, associateParentParam)
+		if err != nil {
+			return errors.Trace(err)
+		}
 	}
 
 	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
-		result := sqlair.M{}
-		err := tx.Query(ctx, queryStmt, machineNameParam).Get(&result)
-		if err != nil {
-			if !errors.Is(err, sqlair.ErrNoRows) {
-				return errors.Annotatef(err, "querying machine %q", machineName)
-			}
-		}
-		// For now, we just care if the minimal machine row already exists.
+		// Query for the machine uuid. If the machine already exists, return a
+		// MachineAlreadyExists error.
+		err := tx.Query(ctx, machineUUIDStmt, machineNameParam).Get(&machineUUIDout)
+		// No error means we found the machine with the given name.
 		if err == nil {
-			return nil
+			return errors.Annotatef(machineerrors.MachineAlreadyExists, "machine %q", mName)
+		} else if !errors.Is(err, sqlair.ErrNoRows) {
```

Comment:
**todo:** `else` not required here, since previous if-branch is a return statement. 

Usually, in golang, we use early return style with no else. I think we should stick on it (and it may broke picky static analysis some day)